extensions/2.0/Khronos/KHR_materials_specular/README.md

```diff
@@ -0,0 +1,244 @@
+
+
+## Extending Materials
+
+The strength of the specular reflection is defined by adding the `KHR_materials_specular` extension to any glTF material.
+
+```json
+{
+    "materials": [
+        {
+            "extensions": {
+                "KHR_materials_specular": {
+                    "specularFactor": 1.0,
+                    "specularColorFactor": [1.0, 1.0, 1.0],
+                    "specularTexture": 0,
+                }
+            }
+        }
+    ]
+}
+```
+
+Factor and texture are combined by multiplication to describe a single value.
+
+| |Type|Description|Required|
+|-|----|-----------|--------|
+| **specularFactor** | `number` | The strength of the specular reflection. | No, default: `1.0`|
+| **specularColorFactor** | `number[3]` | The F0 color of the specular reflection (RGB). | No, default: `[1.0, 1.0, 1.0]`|
+| **specularTexture** | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo) | A 4-channel texture that defines the F0 color of the specular reflection (RGB channels, encoded in sRGB) and the specular factor (A). Will be multiplied by specularFactor and specularColorFactor. | No |
+
+The specular factor scales the microfacet BRDF in the dielectric BRDF. It also affects the diffuse BRDF; the less energy is reflected by the microfacet BRDF, the more can be shifted to the diffuse BRDF. The following image shows specular factor increasing from 0 to 1.
+
+![](figures/specular.png)
+
+The specular color changes the F0 color of the Fresnel that is multiplied to the microfacet BRDF. The color at grazing angles (F90) is not changed. As with specular factor, the diffuse BRDF will be weighted with the directional-dependent remaining energy according to the Fresnel. The weight is an RGB color, involving the complementary to specular color. To make it easy to use and ensure energy conservation, the RGB color is converted to scalar via `max(r, g, b)`. The following images show specular color increasing from [0,0,0] to [1,1,1] (top) and from [0,0,0] to [1,0,0] (bottom).
+
+![](figures/specular-color.png)
+![](figures/specular-color-2.png)
+
+## Implementation
+
+*This section is non-normative.*
+
+The extension changes the computation of the Fresnel term defined in [Appendix B](/specification/2.0/README.md#appendix-b-brdf-implementation) to the following:
+
+```
+dielectricSpecularF0 = 0.04 * specularFactor * specularTexture.a *
+                              specularColorFactor * specularTexture.rgb
+dielectricSpecularF90 = specularFactor * specularTexture.a
+
+F0  = lerp(dielectricSpecularF0, baseColor.rgb, metallic)
+F90 = lerp(dielectricSpecularF90, 1, metallic)
+
+F = F0 + (F90 - F0) * (1 - VdotH)^5
+```
+
+As the Fresnel term `F` is now an RGB value, the diffuse component is also affected:
+
+```
+c_diff = lerp(baseColor.rgb * (1 - max(F0.r, F0.g, F0.b)), black, metallic)
+diffuse = c_diff / PI
+f_diffuse = (1 - max(F.r, F.g, F.b)) * diffuse
```

Comment:
Is there a specific reason why the max value of the fresnel term is taken here? In the [sample viewer](https://github.com/KhronosGroup/glTF-Sample-Viewer/blob/88b18169541ad039efcf370ca168ea69b6b048e3/src/shaders/brdf.glsl#L125) and [Appendix B](https://github.com/DassaultSystemes-Technology/glTF/blob/KHR_materials_specular/specification/2.0/README.md#appendix-b-brdf-implementation) this conversion is not performed.

Reply:
In Appendix B, the Fresnel term for the dielectric case is scalar, so the max is not needed/has no effect.

The full formula for `f_diffuse` from Appendix B:

```
f_diffuse = (1 - F) * diffuse
= (1 - (F0 + (1 - F0) * (1 - VdotH)^5) * diffuse
= (1 - (lerp(dieletricSpecular, baseColor.rgb, metallic) + (1 - lerp(dieletricSpecular, baseColor.rgb, metallic)) * (1 - VdotH)^5) * (1/PI) * lerp(baseColor.rgb * (1 - dielectricSpecular.r), black, metallic)
```

Rewrite/simplify such that metallic is not part of the Fresnel anymore. Then we get a much simpler term for Fresnel `F_dielectric`:

```
f_diffuse = lerp( (1-F_dielectric) * diffuse, black, metallic)

F_dielectric = dielectricSpecular.r + (1 - dielectricSpecular.r) * (1 - VdotH)^5
```

Here it becomes clear that `dielectricSpecular` and `F_dielectric` is scalar in Appendix B. How with KHR_materials_specular we have a color for `dielectricSpecular`, and `F_dielectric` has to be made scalar by using the max value.

Now of course we could compute the `F_dielectric` per-channel instead of taking the max. However I noticed that `KHR_materials_pbrSpecularGlossiness` is also using the max. A quick check in the glTF VSCode plugin showed that Babylon.js and Cesium implement it with max according the extension spec. three.js on the other hand computes the Fresnel independently for each color channel.

Reply:
For some history, see here: https://github.com/KhronosGroup/glTF/issues/1443